bip-0352.mediawiki

```diff
@@ -0,0 +1,472 @@
+
+The one exception is script path spends that use NUMS point ''H'' as their internal key (where ''H = lift_x(0x50929b74c1a04954b78b4b6035e97a5e078a5a0f28ec96d547bfee9ace803ac0)'' which is constructed by taking the hash of the standard uncompressed encoding of the secp256k1 base point ''G'' as X coordinate, see [https://github.com/bitcoin/bips/blob/master/bip-0341.mediawiki#constructing-and-spending-taproot-outputs BIP341: Constructing and spending Taproot outputs] for more details), in which case the output will be skipped for the purposes of shared secret derivation<ref name="why_ignore_h">'''Why skip outputs with H as the internal taproot key?''' If use cases get popularized where the taproot key path cannot be used, these outputs can still be included without getting in the way of making a silent payment, provided they specifically use H as their internal taproot key.</ref>.
+
+''' P2WPKH '''
+
+The sender performs the tweak using the private key for the output and the receiver obtains the public key from the ''witness''.
+
+''' P2SH-P2WPKH '''
+
+The sender performs the tweak using the private key for the nested ''P2WPKH'' output and the receiver obtains the public key from the ''witness''.
+
+''' P2PKH '''
+
+The sender performs the tweak using the private key for the output and SHOULD sign using the standard script template:
+
+
+    scriptSig: <Signature> <Public Key>
+
+
+The receiver obtains the public key from the ''scriptSig''. The receiver MUST parse the ''scriptSig'' for the public key, even if the ''scriptSig'' is non-standard (e.g. <code><dummy> OP_DROP <Signature> <Public Key></code>). This is to address the [https://en.bitcoin.it/wiki/Transaction_malleability third-party malleability of ''P2PKH'' ''scriptSigs''].
+
+''' P2PK '''
+
+The sender performs the tweak using the private key for the output and the receiver obtains the public key from the ''scriptPubKey''.
```

Comment:
Why was P2PK recently added as an allowed input type? Is it really worth the implementation burden [when it's so rarely used](https://transactionfee.info/charts/inputs-and-outputs-p2pk/) and doesn't even have a standard address format?

Reply:
The implementation burden seems very minimal considering it's a simple script and only concerns recipients which already rely on full nodes anyway. In terms of value a significant amount of coins (mostly early miners) is locked up in p2pk.

Reply:
@Sosthene00 I'm afraid that the input scripts need to be checked regardless. Of course it's needed for all taproot inputs (which will already far outweigh the number of p2pk inputs), but perhaps more importantly, it is unsafe to assume you know what the input script will look like based solely on the output script. For instance, the output script may look like p2wpkh but the input script is actually segwit v2 (as of yet undefined, so the script succeeds regardless of what the output script might look like).

I am unsure what exact implementation you had in mind for creating the index, but for Bitcoin Core we are [reading the undo data](https://en.bitcoin.it/wiki/Bitcoin_Core_0.11_(ch_2):_Data_Storage#Raw_undo_data_.28rev.2A.dat.29) to uncover the input scripts for each block. Perhaps you could consider something similar.

Reply:
That's actually an argument for having this index constructed on Bitcoin Core directly, what I've done so far is modifying electrs to construct this index and even if my code is probably rough on the edges and not the most efficient the performance hit is *bad*, a full sync of the signet blockchain takes about 10 times longer than without indexing the silent payments tweaks. And I'm only requesting the scriptpubkey of the prevouts from Core for inputs that spend (or seem to spend) p2tr outputs. 

I agree with you that without requesting the actual scriptpubkey I'm mostly guessing the script type, but requesting it from Core every time would probably make the performance even more abysmal and simply impractical for production. So either :
1. tweaks index is made and served directly from Core
2. we build some ad hoc indexer 
3. there's some other existing indexer out there that is a lot faster than electrs and that could do it (I personnaly won't bet too much on that)

Reply:
@Sosthene00 Yeah that all makes sense to me. It seems hard to make it performant without leaning a bit more on the full node side of things. @Sjors has been exploring [building an index for Core](https://github.com/bitcoin/bitcoin/pull/28241).

Reply:
Yes I was skeptical of burdening Core with that and thought that some implementation of electrum might be a better fit for that job, but now it seems that's a dead end and that we would still be better off doing it in Core

Reply:
I've been going back through and trying to re-write this section with script templates (instead of the the more open ended version we have now), and I think we actually should drop P2PK. Doing so would allow us to get rid of hybrid keys and uncompressed public keys altogether by having our script templates only allow for compressed public keys.

With P2PK, 99% of the value is in uncompressed keys. Furthermore, while there is old value locked, this type of output (and uncompressed keys in general) is almost never used anymore.

For the value that is locked in these coins, if they wanted to use silent payments to move these coins, it would be as simple as adding a single silent payments eligible output to the transaction inputs, as part of a one time move. This seems fine to me since we don't want to encourage using these types of outputs going forward (and the silent payments protocol dictates they would have to move these coins to a P2TR output, anyways).